File: src/parser/rules/expr.c
#include "ast.h"
#include "parser.h"
#include "shell.h"

#include <stdio.h>

// expr: WORD (OPERATOR cmd)*
t_ast *parse_expr(const t_shell *const shell, t_parser *const parser)
{
	t_ast *node = parse_cmd(shell, parser);

	while (IS_VALID_TOKEN(parser->current_token.type, CommandSeparator))
	{
		node = ast_create_node(parser->current_token, node, NULL);
		eat(shell, parser, CommandSeparator);
		node->right = parse_cmd(shell, parser);
	}

	return node;
}

File: src/parser/rules/expr.rs
use crate::ast::Node;
use crate::lexer::TokenType;
use crate::parser::rules::RuleCmd;
use crate::parser::{Parser, ParserError};

const OPERATORS: [TokenType; 4] = [
    TokenType::Pipe,
    TokenType::Redirection,
    TokenType::LogicalAnd,
    TokenType::LogicalOr,
];

fn is_operator(tok_type: &TokenType) -> bool {
    OPERATORS.contains(tok_type)
}

pub trait RuleExpr {
    fn expr(&mut self) -> Result<Node, ParserError>;
}

impl RuleExpr for Parser {
    fn expr(&mut self) -> Result<Node, ParserError> {
        let mut node = self.cmd()?;

        while is_operator(&self.current_token.token_type) {
            let token = self.current_token.clone();
            node = match &token.token_type {
                TokenType::Pipe => {
                    self.eat(TokenType::Pipe)?;
                    Node::new(token, Some(node), Some(self.cmd()?))
                }
                TokenType::Redirection => {
                    self.eat(TokenType::Redirection)?;
                    Node::new(token, Some(node), Some(self.cmd()?))
                }
                TokenType::LogicalAnd => {
                    self.eat(TokenType::LogicalAnd)?;
                    Node::new(token, Some(node), Some(self.cmd()?))
                }
                TokenType::LogicalOr => {
                    self.eat(TokenType::LogicalOr)?;
                    Node::new(token, Some(node), Some(self.cmd()?))
                }
                _ => return Err(ParserError::UnexpectedToken),
            };
        }
        Ok(node)
    }
}
